File: info/1. Алгоритмы/5. Вычислительные алгоритмы/5. Ряд Фибоначчи/code sample/sample3/Sample3.java
package sample3;

import java.math.BigInteger;
import java.util.HashMap;
import java.util.Map;

public class Sample3 {

    static Map<Integer, BigInteger> mem = new HashMap<>();

    public static void main(String[] args) {

        for (int i = 0; i <= 10; i++) {
            System.out.println(i + " -> " + fibonacciSequince(i));
        }
    }

    public static BigInteger fibonacciSequince(int n) {
        BigInteger result = mem.get(n);
        if (result != null) {
            return result;
        } else if (n == 0) {
            mem.put(0, BigInteger.ZERO);
        } else if (n == 1) {
            mem.put(1, BigInteger.ONE);
        } else {
            mem.put(n, fibonacciSequince(n - 1).add(fibonacciSequince(n - 2)));
        }
        return mem.get(n);
    }

}

File: info/1. Алгоритмы/5. Вычислительные алгоритмы/5. Ряд Фибоначчи/code sample/sample_3.py
mem = dict()


def fibonacci_sequince(n):
    if n in mem:
        return mem[n]
    if n == 0:
        mem[n] = 0
    elif n == 1:
        mem[n] = 1
    else:
        mem[n] = fibonacci_sequince(n-1)+fibonacci_sequince(n-2)
    return mem[n]


print(fibonacci_sequince(500))
